WalletWasabi.Tests/UnitTests/Blockchain/TransactionBuilding/BranchAndBoundTests.cs

```diff
@@ -13,81 +14,162 @@ public class BranchAndBoundTests
 	}
 
 	/// <summary>
-	/// Tests that sum of input values must be larger or equal to the target otherwise we end up searching all options in vain.
+	/// Tests that sum of input values must be larger or equal to the target otherwise
+	/// we end up searching all options in vain.
 	/// </summary>
 	[Fact]
-	public void TargetIsBiggerThanBalance()
+	public void ExactMatch_TargetIsBiggerThanBalance()
 	{
-		long target = 5_000;
 		List<long> inputValues = new();
+		long target = 5_000;
 
 		for (int i = 0; i < target - 1; i++)
 		{
 			inputValues.Add(1);
 		}
 
-		BranchAndBound selector = new(inputValues);
-		bool wasSuccessful = selector.TryGetExactMatch(target, out List<long>? selectedCoins);
+		BranchAndBound algorithm = new(inputValues);
+		ExactMatchStrategy strategy = new(target);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
 	}
 
 	[Fact]
-	public void ReturnNullIfNoExactMatchFoundTest()
+	public void ExactMatch_ReturnNullIfNoExactMatchFoundTest()
 	{
 		List<long> inputValues = new() { 120_000, 100_000, 100_000, 50_000, 40_000 };
-		BranchAndBound selector = new(inputValues);
 		long target = 300000;
 
-		bool wasSuccessful = selector.TryGetExactMatch(target, out List<long>? selectedCoins);
+		BranchAndBound algorithm = new(inputValues);
+		ExactMatchStrategy strategy = new(target);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+	}
+
+	/// <summary>Tests that a best found selection is found when an exact solution does not exist.</summary>
+	[Fact]
+	public void CheapestSelection_NoInputCosts()
+	{
+		List<long> inputValues = new() { 11, 7, 5, 3, 2 }; // Sum is 28.
+		long[] inputCosts = new long[] { 0, 0, 0, 0, 0 }; // No input costs. Idealized.
+		long target = 27; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
+		Assert.Equal(new long[] { 11, 7, 5, 3, 2 }, strategy.GetBestSelectionFound());
+	}
+
+	[Fact]
+	public void CheapestSelection_ExactMatchIsAlsoCheapest()
+	{
+		List<long> inputValues = new() { 35, 17, 10, 5, 3, 2 };
+
+		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
+		long[] inputCosts = new long[] { 1, 5, 1, 1, 1, 1 };
+
+		long target = 27;
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+
+		Assert.Equal(new long[] { 17, 10 }, strategy.GetBestSelectionFound());
```

Comment:
This test fails randomly. Sometimes `strategy.GetBestSelectionFound()` can give back:

`Int64[] [ 17, 5, 3, 2 ]` 
and
`Int64[] [ 35 ]`

Something funny is going on.